src/osp/Active/opengl/SysRenderGL.cpp

```diff
@@ -101,134 +105,112 @@ void SysRenderGL::setup_context(RenderGL& rCtxGl)
-                    // get new mesh
-                    rEntMeshGl = try_compile_mesh(rRenderGl, rEntMesh.m_mesh);
-                }
-            }
-            else
-            {
-                // ACompMeshGL component needed
-                rMeshGl.emplace(
-                        ent, try_compile_mesh(rRenderGl, rEntMesh.m_mesh));
-            }
-        }
-        else
+        // New element will be emplaced if it isn't present yet
+        auto const [it, success] = rRenderGl.m_resToMesh.try_emplace(meshRes);
+        if ( ! success)
         {
-            if (rMeshGl.contains(ent))
-            {
-                // ACompMesh removed, remove ACompMeshGL too
-                rMeshGl.erase(ent);
-            }
-            else
-            {
-                // Why is this entity here?
-            }
+            return;
         }
-    }
-}
 
-TexGlId try_compile_texture(
-        RenderGL& rRenderGl, DependRes<ImageData2D> const& texData)
-{
-    auto foundIt = rRenderGl.m_oldResToTex.find(texData.name());
+        // New element emplaced, this means we've just found a resource that
+        // isn't synchronized yet.
 
-    if (foundIt == rRenderGl.m_oldResToTex.end())
-    {
-        // Texture isn't compiled yet, compile it
-        TexGlId newId = rRenderGl.m_texIds.create();
+        // Create new Mesh GL Id
+        MeshGlId const newId = rRenderGl.m_meshIds.create();
 
-        Texture2D &rTexGl = rRenderGl.m_texGl.emplace(newId);
-        
-        using Magnum::GL::SamplerWrapping;
-        using Magnum::GL::SamplerFilter;
-        using Magnum::GL::textureFormat;
+        // Create owner, this adds to the resource's reference count
+        ResIdOwner_t renderOwner
+                = rResources.owner_create(restypes::gc_mesh, meshRes);
 
-        Magnum::ImageView2D view = *texData;
+        // Track with two-way map and store owner
+        rRenderGl.m_meshToRes.emplace(newId, std::move(renderOwner));
+        it->second = newId;
 
-        rTexGl.setWrapping(SamplerWrapping::ClampToEdge)
-            .setMagnificationFilter(SamplerFilter::Nearest)
-            .setMinificationFilter(SamplerFilter::Nearest)
-            .setStorage(1, textureFormat((*texData).format()), (*texData).size())
-            .setSubImage(0, {}, view);
+        // Get mesh data
+        MeshData const &meshData = rResources.data_get<MeshData>(restypes::gc_mesh, meshRes);
 
-        return newId;
+        // Compile and store mesh
+        rRenderGl.m_meshGl.emplace(newId, Magnum::MeshTools::compile(meshData));
     }
-
-    return foundIt->second;
 }
 
-void SysRenderGL::compile_textures(
-        acomp_storage_t<ACompTexture> const& textures,
-        std::vector<ActiveEnt> const& dirty,
-        acomp_storage_t<TexGlId>& rTexGl,
+void SysRenderGL::assign_meshes(
+        acomp_storage_t<MeshIdOwner_t> const& cmpMeshIds,
+        IdMap_t<MeshId, ResIdOwner_t> const& meshToRes,
+        std::vector<ActiveEnt> const& entsDirty,
```

Comment:
I think we can safely start using C++20, all the compilers should be caught up for most C++20 features.

So you could, if you wanted to, use std::span<> for this.

It's basically the sting-view of the non-char world.

Reply:
In this case, Corrade also provides ArrayView equivalent to std::span and is also stl-compatible.

It shouldn't take that much effort to edit the line in the CMakeLists, but switching to C++20 will also affect other people who might want to use the project in the future. Also noting the possibility of using that 'C++ as a scripting language' thing or wasm.

Though I have started using [[likely]] and [[unlikely]] in other projects.

Reply:
the cling interpreter should be able to be built as a shared library for whatever version of c++ it needs, and then it would consume only the public headers of OSP.

If it doesn't support C++20, we can restrict the public headers to C++17, but it's based on clang, so i would hope it's not *too* terribly out of date...

ArrayView is fine, if you'd rather use that.